README.md
# custom_pointed_popup

Popup that can be shown on any targeted widget with customized pointed design.

# Screenshots
![alt text](https://github.com/DipakShrestha-ADS/custom_pointed_popup/blob/master/screenshots/b.png)
![alt text](https://github.com/DipakShrestha-ADS/custom_pointed_popup/blob/master/screenshots/a.png)
![alt text](https://github.com/DipakShrestha-ADS/custom_pointed_popup/blob/master/screenshots/c.png)
![alt text](https://github.com/DipakShrestha-ADS/custom_pointed_popup/blob/master/screenshots/d.png)

## Getting Started
### Example
```dart
  final GlobalKey widgetKey = GlobalKey();

  CustomPointedPopup getCustomPointedPopup(BuildContext context) {
    return CustomPointedPopup(
      backgroundColor: Colors.red,
      context: context,
      lineColor: Colors.tealAccent,
      widthFractionWithRespectToDeviceWidth: 3,
      displayBelowWidget: true,
      triangleDirection: TriangleDirection.FullLeft,
      popupElevation: 10,

      ///you can also add border radius
      ////popupBorderRadius:,
      item: CustomPointedPopupItem(
        title: 'Popup that can be shown on any targeted widget with customized pointed design.',
        textStyle: Theme.of(context).textTheme.caption!.copyWith(
              color: Theme.of(context).cardColor,
            ),
        iconWidget: Icon(
          Icons.add_moderator,
          color: Theme.of(context).cardColor,
        ),

        ///Or you can add custom item widget below instead above 3
        ///itemWidget: Container(),
      ),
      onClickWidget: (onClickMenu) {
        print('popup item clicked');
      },
      onDismiss: () {
        print('on dismissed called');
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(
          'Custom Pointed Popup [CPP]',
        ),
      ),
      body: Container(
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text(
                'CPP with straight pointer & item.',
                style: TextStyle(
                  color: Colors.red,
                  fontSize: 14.0,
                  fontStyle: FontStyle.italic,
                  fontWeight: FontWeight.bold,
                ),
              ),
              SizedBox(
                height: 40,
              ),
              GestureDetector(
                onTap: () {
                  getCustomPointedPopup(context)
                    ..show(
                      widgetKey: widgetKey,
                    );
                },
                child: Card(
                  key: widgetKey,
                  elevation: 10,
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(20),
                  ),
                  child: Container(
                    height: 100,
                    width: 100,
                    child: Center(
                      child: Text(
                        'Click Me \nTo\n Display CPP',
                        textAlign: TextAlign.center,
                        overflow: TextOverflow.clip,
                        style: TextStyle(
                          color: Colors.blueGrey,
                          fontWeight: FontWeight.w900,
                          fontSize: 14.0,
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
```

## Important Class, Function And Enum
The important `class` to know. `CustomPointedPopupItem` - Model item to display inside the pointed popup:

| Attribute | Type  | Default | Required | Description |
|-----------|-------|---------|-------------|----------|
| `iconWidget` | `Widget` |  | `No` | This will create an icon inside the pointed popup. |
| `title` | `String` |  | `No` | Title or text for the pointed popup. |
| `textStyle` | `TextStyle` |  | `No` | Text style for the title. |
| `textAlign` | `TextAlign` |  | `No` | Alignment for the title. |
| `itemWidget` | `Widget` |  | `No` | Custom item widget or child for the pointed popup. No need to add `iconWidget` & `title` if you added this.|

The important function to know. `show` - Function to show the custom pointed popup:

| Attribute | Type  | Default | Required | Description |
|-----------|-------|---------|-------------|----------|
| `rect` | `Rect` |  | `No` | The rect bound that the popup should appear around it. |
| `widgetKey` | `GlobalKey` |  | `No` | The key to widget. |
| `item` | `CustomPointedPopupItem [class]` |  | `No` | If you want the dynamic item widget or child inside the popup per action, you can add the item here. |

The important `enum` to know. `TriangleDirection` - Enum to choose the direction of pointer or triangle shape:

| Value | Description |
|-----------|----------|
| `Straight` | This will create an pointer with straight upward. |
| `Right` | This will create an pointer with slightly bended towards right. |
| `FullRight` | This will create an pointer with fully bended towards right. |
| `Left` | This will create an pointer with slightly bended towards left. |
| `FullLeft` | This will create an pointer with fully bended towards left. |


In order to create a custom pointed popup on the widget, there are several attributes that are supported by `custom_pointed_popup`:

| Attribute | Type  | Default | Required | Description |
|-----------|-------|---------|-------------|----------|
| `item` | `CustomPointedPopupItem [class]` |  | `Yes` | To create the item widget or child in pointed popup. |
| `customHeight` | `double` |  | `Yes` | The custom height of the pointed popup. |
| `displayBelowWidget` | `bool` | `false`  | `No` | If true always display the popup below the widget else automatically calculate the space and display the popup either up or below the widget. |
| `triangleDirection` | `TriangleDirection [enum]` | `TraingleDirection.Straight` | `No` | Determines the direction of pointer or triangle shape. |
| `popupBorderRadius` | `BorderRadius` | `BorderRadius.circular(10.0)` | `No` | Defines the border of the custom pointed popup. |
| `popupElevation` | `double` | `0` | `No` | The elevation of the pointed popup. |
| `stateChanged` | `Function(bool isShow)` |  | `No` | This event function will fire immediately the the popup state changes. |
| `backgroundColor` | `Color` | `Colors.red` | `No` | Background color for pointed popup. |
| `widthFractionWithRespectToDeviceWidth` | `int` | `3` | `No` | Width ratio with respect to the device width. |
| `onDismiss` | `VoidCallBack` |  | `No` | Callback function that fire when the custom pointed popup dismissed. |
| `onClickWidget` | `Function(CustomPointedPopupItem item)` |  | `No` | Callback function that fire when the child of popup is clicked. |
| `context` | `BuildContext` |  | `Yes` | Context where the popup to be shown. |

##If you have any queries, email me to dipak.shrestha@eemc.edu.np

## CREDITS
### Contributors
<a href="https://github.com/DipakShrestha-ADS/custom_pointed_popup/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=DipakShrestha-ADS/custom_pointed_popup" />
</a>

Made with [contributors-img](https://contrib.rocks).


CHANGELOG.md
## 0.0.1 - 2021-07-18

* Popup that can be shown on any targeted widget with customized pointed design.

## 0.0.2 - 2021-07-18

* Documentation added.

## 0.0.3 - 2021-07-18

* Readme updated.

## 0.0.4 - 2021-12-29

* Added custom height for custom popup for both condition of displayBelowWidget boolean.


test/custom_pointed_popup_test.dart
void main() {
  // test('adds one to input values', () {});
}


lib/custom_pointed_popup.dart
library custom_pointed_popup;

import 'dart:core';
import 'dart:ui';

import 'package:custom_pointed_popup/utils/triangle_painter.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

///Custom Pointed Popup Item
class CustomPointedPopupItem {
  Widget? iconWidget;
  String? title;
  TextStyle? textStyle;
  TextAlign? textAlign;

  ///Adding this would replace the popup default item to item widget.
  Widget? itemWidget;

  CustomPointedPopupItem({
    this.title,
    this.iconWidget,
    this.textStyle,
    this.textAlign,
    this.itemWidget,
  }) : assert(title != null || itemWidget != null);

  Widget? get getIconWidget => iconWidget;

  Widget? get getItemWidget => itemWidget;

  String? get getTitle => title;

  TextStyle get getTextStyle => textStyle ?? TextStyle(color: Color(0xffc5c5c5), fontSize: 10.0);

  TextAlign get getTextAlign => textAlign ?? TextAlign.center;
}

typedef WidgetClickCallBack = Function(CustomPointedPopupItem item);
typedef CustomPopupStateChanged = Function(bool isShow);

///Custom Pointed Popup
class CustomPointedPopup {
  static var itemWidth = 72.0;
  static var itemHeight = 65.0;
  static var arrowHeight = 10.0;
  late OverlayEntry _entry;
  late CustomPointedPopupItem item;

  /// row count
  late int _row;

  /// col count
  late int _col;

  /// The left top point of this popup.
  late Offset _offset;

  /// Popup will show at above or under this rect
  late Rect _showRect;

  /// calculated automatically: if false popup is show above of the widget, otherwise popup is show under the widget
  bool _isTriangleDown = true;

  /// The max column count, default is 4.
  late int _widthFractionWithRespectToDeviceWidth;

  /// callback
  late VoidCallback? dismissCallback;
  late WidgetClickCallBack? onClickItemWidget;
  late CustomPopupStateChanged? stateChanged;

  late Size _screenSize;

  /// Cannot be null
  static late BuildContext context;

  /// style
  late Color _backgroundColor;

  /// It's showing or not.
  bool _isShow = false;
  bool get isShow => _isShow;

  ///always displayed popup below the widget.
  bool displayBelowWidget;
  double? customHeight;

  TriangleDirection? triangleDirection;

  BorderRadius? popupBorderRadius;

  double? popupElevation;

  CustomPointedPopup({
    WidgetClickCallBack? onClickWidget,
    required BuildContext context,
    VoidCallback? onDismiss,
    int? widthFractionWithRespectToDeviceWidth,
    Color? backgroundColor,
    CustomPopupStateChanged? stateChanged,
    required CustomPointedPopupItem item,
    this.displayBelowWidget = false,
    this.triangleDirection,
    this.popupBorderRadius,
    this.popupElevation,
    this.customHeight,
  }) {
    this.onClickItemWidget = onClickWidget;
    this.dismissCallback = onDismiss;
    this.stateChanged = stateChanged;
    this.item = item;
    this._widthFractionWithRespectToDeviceWidth = widthFractionWithRespectToDeviceWidth ?? 3;
    this._backgroundColor = backgroundColor ?? Colors.red;
    CustomPointedPopup.context = context;
  }

  void show({Rect? rect, GlobalKey? widgetKey, CustomPointedPopupItem? item}) {
    assert(rect != null || widgetKey != null);
    if (rect == null && widgetKey == null) {
      throw ErrorWidget.withDetails(
        message: 'Both Rect & Widget key can\'t be null.',
      );
    }

    this.item = item ?? this.item;
    this._showRect = rect ?? CustomPointedPopup.getWidgetGlobalRect(widgetKey!);
    this._screenSize = window.physicalSize / window.devicePixelRatio;
    this.dismissCallback = dismissCallback;

    _calculatePosition(CustomPointedPopup.context);

    _entry = OverlayEntry(builder: (context) {
      return buildPopupMenuLayout(_offset);
    });

    Overlay.of(CustomPointedPopup.context)!.insert(_entry);
    _isShow = true;
    if (this.stateChanged != null) {
      this.stateChanged!(true);
    }
  }

  static Rect getWidgetGlobalRect(GlobalKey key) {
    RenderBox renderBox = key.currentContext!.findRenderObject()! as RenderBox;
    var offset = renderBox.localToGlobal(Offset.zero);
    return Rect.fromLTWH(offset.dx, offset.dy, renderBox.size.width, renderBox.size.height);
  }

  void _calculatePosition(BuildContext context) {
    _col = _calculateColCount();
    _row = _calculateRowCount();
    _offset = _calculateOffset(CustomPointedPopup.context);
  }

  Offset _calculateOffset(BuildContext context) {
    double dx = _showRect.left + _showRect.width / 2.0 - popupWidth() / 2.0;
    if (dx < 10.0) {
      dx = 10.0;
    }

    if (dx + popupWidth() > _screenSize.width && dx > 10.0) {
      double tempDx = _screenSize.width - popupWidth() - 10;
      if (tempDx > 10) dx = tempDx;
    }

    double dy = _showRect.top - (customHeight ?? popupHeight());

    if (displayBelowWidget) {
      dy = arrowHeight + _showRect.height + _showRect.top;
      _isTriangleDown = false;
    } else {
      if (dy <= MediaQuery.of(context).padding.top + 10) {
        dy = arrowHeight + _showRect.height + _showRect.top;
        _isTriangleDown = false;
      } else {
        dy -= arrowHeight;
        _isTriangleDown = true;
      }
    }

    return Offset(dx, dy);
  }

  double popupWidth() {
    return itemWidth * _col;
  }

  double popupHeight() {
    var height = itemHeight * _row;
    return height;
  }

  LayoutBuilder buildPopupMenuLayout(Offset offset) {
    return LayoutBuilder(builder: (context, constraints) {
      return GestureDetector(
        behavior: HitTestBehavior.translucent,
        onTap: () {
          dismiss();
        },
        onVerticalDragStart: (DragStartDetails details) {
          dismiss();
        },
        onHorizontalDragStart: (DragStartDetails details) {
          dismiss();
        },
        child: Container(
          child: Stack(
            children: <Widget>[
              // Triangle arrow paint
              Positioned(
                left: _showRect.left + _showRect.width / 2.0 - 7.5,
                top: _isTriangleDown ? offset.dy + (customHeight ?? popupHeight()) : (offset.dy - arrowHeight),
                child: CustomPaint(
                  size: Size(15.0, arrowHeight),
                  painter: TrianglePainter(
                    isDown: _isTriangleDown,
                    color: _backgroundColor,
                    triangleDirection: triangleDirection != null ? triangleDirection! : TriangleDirection.Straight,
                  ),
                ),
              ),
              // Widgets
              Positioned(
                left: offset.dx,
                top: offset.dy,
                child: Card(
                  elevation: popupElevation ?? 0,
                  margin: EdgeInsets.zero,
                  shape: RoundedRectangleBorder(
                    borderRadius: popupBorderRadius ?? BorderRadius.circular(10.0),
                  ),
                  child: Container(
                    width: popupWidth(),
                    height: displayBelowWidget ? customHeight : (customHeight ?? popupHeight()),
                    child: Column(
                      children: <Widget>[
                        ClipRRect(
                          borderRadius: popupBorderRadius ?? BorderRadius.circular(10.0),
                          child: Container(
                            width: popupWidth(),
                            height: displayBelowWidget ? customHeight : (customHeight ?? popupHeight()),
                            padding: EdgeInsets.all(5.0),
                            decoration: BoxDecoration(
                              color: _backgroundColor,
                              borderRadius: popupBorderRadius ?? BorderRadius.circular(10.0),
                            ),
                            child: GestureDetector(
                              onTap: () {
                                itemClicked(item);
                              },
                              child: Column(
                                mainAxisAlignment: MainAxisAlignment.center,
                                children: item.itemWidget != null
                                    ? [item.itemWidget!]
                                    : [
                                        item.getIconWidget!,
                                        Text(
                                          '${item.getTitle}',
                                          style: item.getTextStyle,
                                          textAlign: item.getTextAlign,
                                          overflow: displayBelowWidget ? TextOverflow.clip : TextOverflow.ellipsis,
                                        ),
                                      ],
                              ),
                            ),
                          ),
                        ),
                      ],
                    ),
                  ),
                ),
              )
            ],
          ),
        ),
      );
    });
  }

  int _calculateRowCount() {
    int itemCount = 1;

    if (_calculateColCount() == 1) {
      return itemCount;
    }

    int row = (itemCount - 1) ~/ _calculateColCount() + 1;

    return row;
  }

  int _calculateColCount() {
    if (_widthFractionWithRespectToDeviceWidth >= 5) {
      return 4;
    }

    return _widthFractionWithRespectToDeviceWidth;
  }

  double get screenWidth {
    double width = window.physicalSize.width;
    double ratio = window.devicePixelRatio;
    return width / ratio;
  }

  void itemClicked(CustomPointedPopupItem item) {
    if (onClickItemWidget != null) {
      onClickItemWidget!(item);
    }

    dismiss();
  }

  void dismiss() {
    if (!_isShow) {
      return;
    }

    _entry.remove();
    _isShow = false;
    if (dismissCallback != null) {
      dismissCallback!();
    }

    if (this.stateChanged != null) {
      this.stateChanged!(false);
    }
  }
}


lib/utils/triangle_painter.dart
import 'package:flutter/rendering.dart';

class TrianglePainter extends CustomPainter {
  bool isDown;
  Color color;
  TriangleDirection triangleDirection;
  TrianglePainter({
    this.isDown = true,
    required this.color,
    this.triangleDirection = TriangleDirection.Left,
  });

  @override
  void paint(Canvas canvas, Size size) {
    Paint _paint = new Paint();
    _paint.strokeWidth = 2.0;
    _paint.color = color;
    _paint.style = PaintingStyle.fill;

    Path path = new Path();
    if (isDown) {
      path.moveTo(0.0, -1.0);
      path.lineTo(size.width, -1.0);
      if (triangleDirection == TriangleDirection.Straight) {
        path.lineTo(size.width / 2.0, size.height);
      } else if (triangleDirection == TriangleDirection.FullLeft) {
        path.lineTo(size.width / -2.0, size.height);
      } else if (triangleDirection == TriangleDirection.Left) {
        path.lineTo(size.width / 4.0, size.height);
      } else if (triangleDirection == TriangleDirection.FullRight) {
        path.lineTo(size.width / 0.5, size.height);
      } else if (triangleDirection == TriangleDirection.Right) {
        path.lineTo(size.width / 1.0, size.height);
      }
    } else {
      if (triangleDirection == TriangleDirection.Straight) {
        path.moveTo(size.width / 2.0, 0.0);
      } else if (triangleDirection == TriangleDirection.FullLeft) {
        path.moveTo(size.width / -2.0, 0.0);
      } else if (triangleDirection == TriangleDirection.Left) {
        path.moveTo(size.width / 4.0, 0.0);
      } else if (triangleDirection == TriangleDirection.FullRight) {
        path.moveTo(size.width / 0.5, 0.0);
      } else if (triangleDirection == TriangleDirection.Right) {
        path.moveTo(size.width / 1.0, 0.0);
      }

      path.lineTo(0.0, size.height + 1);
      path.lineTo(size.width, size.height + 1);
    }

    canvas.drawPath(path, _paint);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}

enum TriangleDirection {
  Straight,
  Right,
  FullRight,
  Left,
  FullLeft,
}


example/lib/main.dart
import 'package:custom_pointed_popup/custom_pointed_popup.dart';
import 'package:custom_pointed_popup/utils/triangle_painter.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blueGrey,
      ),
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  final GlobalKey widgetKey = GlobalKey();

  CustomPointedPopup getCustomPointedPopup(BuildContext context) {
    return CustomPointedPopup(
      backgroundColor: Colors.red,
      context: context,
      widthFractionWithRespectToDeviceWidth: 3,
      triangleDirection: TriangleDirection.Straight,
      popupElevation: 10,
      displayBelowWidget: false,
      // customHeight: 150,

      ///you can also add border radius
      ////popupBorderRadius:,
      item: CustomPointedPopupItem(
        title: 'Popup that can be shown on any targeted widget with customized pointed design.',
        textStyle: Theme.of(context).textTheme.caption!.copyWith(
              color: Theme.of(context).cardColor,
            ),
        iconWidget: Icon(
          Icons.add_moderator,
          color: Theme.of(context).cardColor,
        ),

        ///Or you can add custom item widget below instead above 3
        // itemWidget: Container(
        //   child: Center(
        //     child: Text('CUstome item'),
        //   ),
        // ),
      ),
      onClickWidget: (onClickMenu) {
        print('popup item clicked');
      },
      onDismiss: () {
        print('on dismissed called');
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(
          'Custom Pointed Popup [CPP]',
        ),
      ),
      body: Container(
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text(
                'CPP with straight pointer & item.',
                style: TextStyle(
                  color: Colors.red,
                  fontSize: 14.0,
                  fontStyle: FontStyle.italic,
                  fontWeight: FontWeight.bold,
                ),
              ),
              SizedBox(
                height: 40,
              ),
              GestureDetector(
                onTap: () {
                  getCustomPointedPopup(context)
                    ..show(
                      widgetKey: widgetKey,
                    );
                },
                child: Card(
                  key: widgetKey,
                  elevation: 10,
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(20),
                  ),
                  child: Container(
                    height: 100,
                    width: 100,
                    child: Center(
                      child: Text(
                        'Click Me \nTo\n Display CPP',
                        textAlign: TextAlign.center,
                        overflow: TextOverflow.clip,
                        style: TextStyle(
                          color: Colors.blueGrey,
                          fontWeight: FontWeight.w900,
                          fontSize: 14.0,
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
